api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsWhoPickeupArvDuringPeriodDataDefinitionQueries.java

```diff
@@ -80,109 +80,192 @@ public DataDefinition getLastStateOrDateFromARTProgramOrFichaResumoOrClinica(boo
     } else {
-      query +=
-          " SELECT p.patient_id, CASE "
-              + "   WHEN  e.encounter_type = ${53} THEN o.obs_datetime   "
-              + "   WHEN e.encounter_type = ${6} THEN e.encounter_datetime   "
-              + "   END AS state_date ";
+      query =
+          " SELECT patient_id, "
+              + " Max(state_date) "
+              + " FROM (SELECT p.patient_id AS patient_id, "
+              + " e.encounter_datetime AS state_date, "
+              + " o.value_coded AS state "
+              + " FROM   patient p "
+              + " INNER JOIN encounter e "
+              + " ON p.patient_id = e.patient_id "
+              + " INNER JOIN obs o "
+              + " ON e.encounter_id = o.encounter_id "
+              + " WHERE  p.voided = 0 "
+              + " AND e.voided = 0 "
+              + " AND o.voided = 0 "
+              + " AND e.encounter_type = ${6} "
+              + " AND e.location_id = :location "
+              + " AND e.encounter_datetime <= :endDate "
+              + " AND o.concept_id = ${6272} "
+              + " AND o.value_coded IN( ${1705}, ${1706}, ${1709}, ${1707}, ${1366}, ${23903}, ${1369} ) "
+              + " UNION "
+              + " SELECT p.patient_id   AS patient_id, "
+              + " o.obs_datetime AS state_date, "
+              + " o.value_coded  AS state "
+              + " FROM   patient p "
+              + " INNER JOIN encounter e "
+              + " ON p.patient_id = e.patient_id "
+              + " INNER JOIN obs o "
+              + " ON e.encounter_id = o.encounter_id "
+              + " WHERE  p.voided = 0 "
+              + " AND e.voided = 0 "
+              + " AND o.voided = 0 "
+              + " AND e.encounter_type = ${53} "
+              + " AND e.location_id = :location "
+              + " AND e.encounter_datetime <= :endDate "
+              + " AND o.concept_id = ${6272} "
+              + " AND o.value_coded IN( ${1705}, ${1706}, ${1709}, ${1707}, ${1366}, ${23903}, ${1369} ) "
+              + " UNION "
+              + " SELECT p.patient_id  AS patient_id, "
+              + " ps.start_date AS state_date, "
+              + " ps.state AS state "
+              + " FROM   patient p "
+              + " INNER JOIN patient_program pg "
+              + " ON p.patient_id = pg.patient_id "
+              + " INNER JOIN program pr "
+              + " ON pr.program_id = pg.program_id "
+              + " INNER JOIN patient_state ps "
+              + " ON pg.patient_program_id = ps.patient_program_id "
+              + " WHERE  ps.state IN( ${state_6}, ${7}, ${8}, ${9}, ${10}, ${29} ) "
+              + " AND pg.location_id = :location "
+              + " AND ps.start_date <= :endDate "
+              + " AND p.voided = 0 "
+              + " AND pg.voided = 0 "
+              + " AND pr.retired = 0 "
+              + " AND ps.voided = 0) summary "
+              + " GROUP  BY patient_id ";
```

Comment:
Should format the query to improve the readability

Reply:
This query is no longer on the code, it was re-implemented on the last fix